# Lab on Pandas
## material was originally design for CMSC 12100/CAPP 30121.

In [131]:
import pandas as pd
import numpy as np

### Reading the data
We will be using a sample dataset from the Current Population Survey for this assignment. The file morg_d07_strings.csv contains a modified version of the 2007 MORG data, which we downloaded from the Census Bureau’s website.

The file is in comma-separated value (CSV) format. It can be understood to represent a table with multiple rows and columns (in fact, the CSV format is supported by most spreadsheet programs, and you can try opening the file in Excel, Libreoffice Calc, etc.). The first line of the file is the header of the file. It contains the names of the columns, separated by commas. After the header, each line in the file represents a row in the table, with each value in the row (corresponding to the columns specified in the header) separated by a comma. A common way to refer to a value in a row is as a field. So, if a CSV file has an age column, in an individual row we would refer to the age field (instead of column, which tends to refer to an entire column of values).

Each row in morg_d07_strings.csv corresponds to the survey data obtained from a unique individual. We consider the following variables for each individual in this assignment (although there are a lot more variables available in the MORG datasets):

- h_id: a string that serves as a unique identifier, which we created by concatenating several variables in the original MORG datasets.
- age: an integer value specifying the age of the individual.
- gender: the gender (or sex) recorded for the individual.
- race: the race recorded for the individual.
- ethnicity: the ethnicity recorded for the individual.
- employment_status: the employment status record for the individual.
- hours_worked_per_week: an integer that specifies the usual weekly work hours of the individual.
- earnings_per_week: a float that indicates the weekly earnings of the individual.
The CSV file has a column for each of these variables. Here are the first few lines of the file:

In [132]:
morg_df = pd.read_csv('data/morg_d07_strings.csv')
morg_df.head(10)

h_id  age  gender       race     ethnicity employment_status  \
0    1_1_1   32  Female  BlackOnly  Non-Hispanic           Working   
1    1_2_2   80  Female  WhiteOnly  Non-Hispanic           Others2   
2    1_3_3   20  Female  BlackOnly  Non-Hispanic           Others2   
3    1_4_4   28    Male  WhiteOnly  Non-Hispanic           Working   
4    1_5_5   32    Male  WhiteOnly  Non-Hispanic           Working   
5    1_6_6   69  Female  WhiteOnly  Non-Hispanic           Others1   
6    1_7_7   80  Female  WhiteOnly  Non-Hispanic           Others1   
7    1_8_8   31    Male  WhiteOnly  Non-Hispanic           Working   
8    1_9_9   68  Female  WhiteOnly  Non-Hispanic           Working   
9  1_11_11   75    Male  WhiteOnly  Non-Hispanic           Others1   

   hours_worked_per_week  earnings_per_week  
0                   40.0            1250.00  
1                    NaN                NaN  
2                    NaN                NaN  
3                   40.0            1100.00  
4                   52.0            1289.23  
5                    NaN                NaN  
6                    NaN                NaN  
7                   45.0             866.25  
8                   10.0             105.00  
9                    NaN                NaN

#### Task 1: 
Use pd.read_csv to read the sample data into a pandas dataframe and save the result in a variable named morg_df. Use h_id, which uniquely identifies each row, and as the row index.

In [133]:
morg_df = pd.read_csv('data/morg_d07_strings.csv', index_col = 'h_id')
morg_df.head()

age  gender       race     ethnicity employment_status  \
h_id                                                            
1_1_1   32  Female  BlackOnly  Non-Hispanic           Working   
1_2_2   80  Female  WhiteOnly  Non-Hispanic           Others2   
1_3_3   20  Female  BlackOnly  Non-Hispanic           Others2   
1_4_4   28    Male  WhiteOnly  Non-Hispanic           Working   
1_5_5   32    Male  WhiteOnly  Non-Hispanic           Working   

       hours_worked_per_week  earnings_per_week  
h_id                                             
1_1_1                   40.0            1250.00  
1_2_2                    NaN                NaN  
1_3_3                    NaN                NaN  
1_4_4                   40.0            1100.00  
1_5_5                   52.0            1289.23

### Some simple analysis
Use .dtypes, .shape and .describe()

In [134]:
morg_df.dtypes

age                        int64
gender                    object
race                      object
ethnicity                 object
employment_status         object
hours_worked_per_week    float64
earnings_per_week        float64
dtype: object

In [135]:
morg_df.shape

(284020, 7)

In [136]:
morg_df.size

1988140

In [137]:
len(morg_df)

284020

In [138]:
morg_df.describe()

age  hours_worked_per_week  earnings_per_week
count  284020.000000          166620.000000      166315.000000
mean       45.594757              38.773695         767.404918
std        18.475112              10.462456         566.180836
min        16.000000               0.000000           0.000000
25%        30.000000              40.000000         384.610000
50%        45.000000              40.000000         615.380000
75%        59.000000              40.000000         990.380000
max        85.000000              99.000000        2884.610000

In [139]:
morg_df.columns

Index(['age', 'gender', 'race', 'ethnicity', 'employment_status',
       'hours_worked_per_week', 'earnings_per_week'],
      dtype='object')

## Getting values

#### Task 2: 
Extract the "age" column from morg_df.

In [140]:
morg_df['age'].head()

h_id
1_1_1    32
1_2_2    80
1_3_3    20
1_4_4    28
1_5_5    32
Name: age, dtype: int64

#### Task 3: 
Extract the row that corresponds to h_id 1_2_2 from morg_df.

In [141]:
morg_df.loc['1_2_2',]

age                                80
gender                         Female
race                        WhiteOnly
ethnicity                Non-Hispanic
employment_status             Others2
hours_worked_per_week             NaN
earnings_per_week                 NaN
Name: 1_2_2, dtype: object

In [142]:
type(morg_df.loc['1_2_2',])

pandas.core.series.Series

In [143]:
morg_df.loc['1_2_2', 'gender']

'Female'

In [144]:
morg_df.iloc[3,]

age                                28
gender                           Male
race                        WhiteOnly
ethnicity                Non-Hispanic
employment_status             Working
hours_worked_per_week              40
earnings_per_week                1100
Name: 1_4_4, dtype: object

#### Task 4: 
Use slicing to extract the first four rows of morg_df.

In [145]:
morg_df.iloc[:4]

age  gender       race     ethnicity employment_status  \
h_id                                                            
1_1_1   32  Female  BlackOnly  Non-Hispanic           Working   
1_2_2   80  Female  WhiteOnly  Non-Hispanic           Others2   
1_3_3   20  Female  BlackOnly  Non-Hispanic           Others2   
1_4_4   28    Male  WhiteOnly  Non-Hispanic           Working   

       hours_worked_per_week  earnings_per_week  
h_id                                             
1_1_1                   40.0             1250.0  
1_2_2                    NaN                NaN  
1_3_3                    NaN                NaN  
1_4_4                   40.0             1100.0

#### Task 5:
Use:
- .isna()
- .isna().all()
- .isna().any(axis=1)
- .isna().any(axis=0)
See what is the difference

In [146]:
morg_df.isna().head()

age  gender   race  ethnicity  employment_status  \
h_id                                                        
1_1_1  False   False  False      False              False   
1_2_2  False   False  False      False              False   
1_3_3  False   False  False      False              False   
1_4_4  False   False  False      False              False   
1_5_5  False   False  False      False              False   

       hours_worked_per_week  earnings_per_week  
h_id                                             
1_1_1                  False              False  
1_2_2                   True               True  
1_3_3                   True               True  
1_4_4                  False              False  
1_5_5                  False              False

In [147]:
morg_df.isna().any().all()

False

In [148]:
morg_df.isna().any(axis = 1).head()

h_id
1_1_1    False
1_2_2     True
1_3_3     True
1_4_4    False
1_5_5    False
dtype: bool

In [149]:
morg_df.isna().all()

age                      False
gender                   False
race                     False
ethnicity                False
employment_status        False
hours_worked_per_week    False
earnings_per_week        False
dtype: bool

### Task 6: 
Replace the NA values. Use the fillna method to replace the missing values in the columns you identified in the previous task with zero.

In [150]:
morg_df.fillna(0, inplace = True)

In [151]:
morg_df.isna().any()

age                      False
gender                   False
race                     False
ethnicity                False
employment_status        False
hours_worked_per_week    False
earnings_per_week        False
dtype: bool

### Task 9: 
Use filtering to extract all rows that correspond to a person who works 35 or more hours per week.

In [152]:
morg_df[morg_df['hours_worked_per_week'] >= 35].head()

age  gender       race     ethnicity employment_status  \
h_id                                                              
1_1_1     32  Female  BlackOnly  Non-Hispanic           Working   
1_4_4     28    Male  WhiteOnly  Non-Hispanic           Working   
1_5_5     32    Male  WhiteOnly  Non-Hispanic           Working   
1_8_8     31    Male  WhiteOnly  Non-Hispanic           Working   
1_14_14   19  Female  BlackOnly  Non-Hispanic           Working   

         hours_worked_per_week  earnings_per_week  
h_id                                               
1_1_1                     40.0            1250.00  
1_4_4                     40.0            1100.00  
1_5_5                     52.0            1289.23  
1_8_8                     45.0             866.25  
1_14_14                   40.0             350.00

### Task 7:
Use filtering to extract the rows that correspond to the people who are not working.

In [153]:
morg_df[morg_df['hours_worked_per_week'] == 0].head()

age  gender       race     ethnicity employment_status  \
h_id                                                              
1_2_2     80  Female  WhiteOnly  Non-Hispanic           Others2   
1_3_3     20  Female  BlackOnly  Non-Hispanic           Others2   
1_6_6     69  Female  WhiteOnly  Non-Hispanic           Others1   
1_7_7     80  Female  WhiteOnly  Non-Hispanic           Others1   
1_11_11   75    Male  WhiteOnly  Non-Hispanic           Others1   

         hours_worked_per_week  earnings_per_week  
h_id                                               
1_2_2                      0.0                0.0  
1_3_3                      0.0                0.0  
1_6_6                      0.0                0.0  
1_7_7                      0.0                0.0  
1_11_11                    0.0                0.0

### Task 8:
Use filtering to extract the rows that correspond to people who worked at least 35 hours per week or who earned more than $1000 per week.

In [154]:
morg_df[(morg_df['hours_worked_per_week'] >= 35) & (morg_df['earnings_per_week'] > 1000)].head()

age  gender       race     ethnicity employment_status  \
h_id                                                              
1_1_1     32  Female  BlackOnly  Non-Hispanic           Working   
1_4_4     28    Male  WhiteOnly  Non-Hispanic           Working   
1_5_5     32    Male  WhiteOnly  Non-Hispanic           Working   
1_17_17   55  Female  WhiteOnly  Non-Hispanic           Working   
1_51_51   31    Male  WhiteOnly  Non-Hispanic           Working   

         hours_worked_per_week  earnings_per_week  
h_id                                               
1_1_1                     40.0            1250.00  
1_4_4                     40.0            1100.00  
1_5_5                     52.0            1289.23  
1_17_17                   45.0            1153.84  
1_51_51                   45.0            1400.00

## Task 9:
Create a new DataFrame with people that worked. call it morg_df_worked

In [155]:
morg_df_worked = morg_df[morg_df['hours_worked_per_week'] > 0].copy()
#.copy() helps us tell pandas that we are in fact making a copy of
#a DataFrame. Without copy, we get the warning in the next step. 
#Pandas is warning us that we only make changes in the new dataframe when
#we may have wanted to do them in the original one.

### Task 10:
Using mord_df_worked, create a new column with average earnings per hour called "avg_earnings_per_hour"

In [156]:
morg_df_worked.loc['avg_earnings_per_hr'] = morg_df_worked['earnings_per_week'] / morg_df_worked['hours_worked_per_week']

### Task 12:
Create a new column that has the following string: "Earnings per week are: <earnings_per_week>" using map. Call it 'string_earnings'.

In [157]:
morg_df_worked['string_earnings'] = morg_df_worked['earnings_per_week'].map("Earnings per week are {}".format)
morg_df_worked['string_earnings'].head()

h_id
1_1_1     Earnings per week are 1250.0
1_4_4     Earnings per week are 1100.0
1_5_5    Earnings per week are 1289.23
1_8_8     Earnings per week are 866.25
1_9_9      Earnings per week are 105.0
Name: string_earnings, dtype: object

### Task 13:

Create a new dummy column 'female that has value of 1 if female. Use .map with a dictionary like this:
gender_dict = {'Female': 1,
               'Male': 0}

In [158]:
gender_dict = {'Female': 1, 'Male': 0}
morg_df_worked['female'] = morg_df_worked['gender'].map(gender_dict)
morg_df_worked['female'].head()

h_id
1_1_1    1.0
1_4_4    0.0
1_5_5    0.0
1_8_8    0.0
1_9_9    1.0
Name: female, dtype: float64

In [159]:
morg_df_worked['female'].unique()

array([ 1.,  0., nan])

### Task 14:
Create a new column 'hard_worker' equal to 1 if person worked more than the average hours worked. Use **apply**


In [160]:
mean_hours_worked = morg_df_worked['hours_worked_per_week'].mean()
def hard_worker(hours_worked):
    if hours_worked > mean_hours_worked:
        return 1
    else:
        return 0
morg_df_worked['person_hard_worker'] = morg_df_worked['hours_worked_per_week'].apply(hard_worker)
morg_df_worked[['hours_worked_per_week', 'person_hard_worker']].head(10)

hours_worked_per_week  person_hard_worker
h_id                                              
1_1_1                     40.0                   1
1_4_4                     40.0                   1
1_5_5                     52.0                   1
1_8_8                     45.0                   1
1_9_9                     10.0                   0
1_13_13                   30.0                   0
1_14_14                   40.0                   1
1_17_17                   45.0                   1
1_18_18                   21.0                   0
1_19_19                   40.0                   1

### Task 15:
Use **apply** to calculate the max of earnings_per_week and hours_worked_per_week

In [161]:
morg_df_worked['person_hard_worker'] = \
    morg_df_worked['hours_worked_per_week'].apply(lambda x: 1 if x > mean_hours_worked else 0)

### Task 16:
Create a function that takes a string and reverses it. Use apply to create a new column 'reversed' with employment_status reversed.

In [162]:
def reverse(string):
    return string[::-1]


## Merging

In [163]:
students = pd.read_csv("data/students.csv")
grades = pd.read_csv("data/grades.csv")

In [164]:
students

First Name   Last Name  UCID                     Email        Major
0        Sam       Spade  1234        spade@uchicago.edu    Sociology
1      Nancy        Drew  2789        ndrew@uchicago.edu  Mathematics
2   Sherlock      Holmes  2222  bakerstreet@uchicago.edu   Psychology
3       V.I.  Warshawski  7654          viw@uchicago.edu  Mathematics

In [165]:
grades

Course  UCID  Score Grade
0  CS 121  2789     90    A-
1  CS 121  1234     65     C
2  CS 121  7654     85    B+
3  CS 121  9999    100    A+

In [166]:
pd.merge(students, grades, on="UCID", how="inner")

First Name   Last Name  UCID               Email        Major  Course  \
0        Sam       Spade  1234  spade@uchicago.edu    Sociology  CS 121   
1      Nancy        Drew  2789  ndrew@uchicago.edu  Mathematics  CS 121   
2       V.I.  Warshawski  7654    viw@uchicago.edu  Mathematics  CS 121   

   Score Grade  
0     65     C  
1     90    A-  
2     85    B+

Notice that the columns from the first argument (students) are followed by the corresponding columns from the second argument (grades) minus the UCID and that each row contains information for the same UCID. Sherlock Holmes does not appear in the result, because there is no row with his UCID (2222) in the grades `` dataframe.  Also, notice that UCID 9999, which appears in the ``grades dataframe, does not appear in the result, because it has no mate in the students dataframe.

### Task 17 
Include Sherlock, dont include UCID 9999

In [167]:
pd.merge(students, grades, on="UCID", how="left")

First Name   Last Name  UCID                     Email        Major  Course  \
0        Sam       Spade  1234        spade@uchicago.edu    Sociology  CS 121   
1      Nancy        Drew  2789        ndrew@uchicago.edu  Mathematics  CS 121   
2   Sherlock      Holmes  2222  bakerstreet@uchicago.edu   Psychology     NaN   
3       V.I.  Warshawski  7654          viw@uchicago.edu  Mathematics  CS 121   

   Score Grade  
0   65.0     C  
1   90.0    A-  
2    NaN   NaN  
3   85.0    B+

### Task 18
Include UCID 9999, dont include Sherlock

In [168]:
pd.merge(students, grades, on="UCID", how="right")

First Name   Last Name  UCID               Email        Major  Course  \
0        Sam       Spade  1234  spade@uchicago.edu    Sociology  CS 121   
1      Nancy        Drew  2789  ndrew@uchicago.edu  Mathematics  CS 121   
2       V.I.  Warshawski  7654    viw@uchicago.edu  Mathematics  CS 121   
3        NaN         NaN  9999                 NaN          NaN  CS 121   

   Score Grade  
0     65     C  
1     90    A-  
2     85    B+  
3    100    A+

### Task 19
Include all

In [169]:
pd.merge(students, grades, on="UCID", how="outer")

First Name   Last Name  UCID                     Email        Major  Course  \
0        Sam       Spade  1234        spade@uchicago.edu    Sociology  CS 121   
1      Nancy        Drew  2789        ndrew@uchicago.edu  Mathematics  CS 121   
2   Sherlock      Holmes  2222  bakerstreet@uchicago.edu   Psychology     NaN   
3       V.I.  Warshawski  7654          viw@uchicago.edu  Mathematics  CS 121   
4        NaN         NaN  9999                       NaN          NaN  CS 121   

   Score Grade  
0   65.0     C  
1   90.0    A-  
2    NaN   NaN  
3   85.0    B+  
4  100.0    A+